misc/shunting_yard.c

```diff
@@ -0,0 +1,225 @@
+#include <ctype.h>      /// for isdigit()
+
+/**
+ * @brief Helper function that returns each operator's precedence
+ * @param operator the operator to be queried
+ * @returns the operator's precedence
+ */
+int getPrecedence(char operator) {
+    switch (operator) {
+        case '+':
+        case '-': {
+            return 1;
+        }
+        case '*':
+        case '/': {
+            return 2;
+        }
+        case '^': {
+            return 3;
+        }
+        default:{
+            fprintf(stderr,"Error: Invalid operator\n");
+            return -1;
+        }
+    }
+}
+
+/**
+ * @brief Helper function that returns each operator's associativity
+ * @param operator the operator to be queried
+ * @returns '1' if the operator is left associative
+ * @returns '0' if the operator is right associative
+ */
+int getAssociativity(char operator) {
+    switch (operator) {
+        case '^': {
+            return 0;
+        }
+        case '+':
+        case '-':
+        case '*':
+        case '/': {
+            return 1;
+        }
+        default: {
+            fprintf(stderr,"Error: Invalid operator\n");
+            return -1;
+        }
+    }
+}
+
+/**
+ * @brief An implementation of the shunting yard that converts infix notation to reversed polish notation
+ * @param input pointer to input string
+ * @param output pointer to output location
+ * @returns `1` if a parentheses mismatch is detected
+ * @returns `0` if no mismatches are detected
+ */
+int shuntingYard(const char *input, char *output) {
+    const unsigned int inputLength = strlen(input);
+    char* operatorStack = (char*) malloc(sizeof(char) * inputLength);
+
+    // This pointer points at where we should insert the next element,
+    // Hence stackPointer - 1 is used when accessing elements
+    unsigned int stackPointer = 0;
+
+    // We will parse the input with strtok(),
+    // Since strtok() is destructive, we make a copy of the input to preserve the original string
+    char* str = malloc(sizeof(char) * inputLength + 1);
+    strcpy(str,input);
+    char* token = strtok(str," ");
+
+    // We will push to output with strcat() and strncat(),
+    // This initializes output to be a string with a length of zero
+    output[0] = '\0';
+
+    while (token != NULL) {
+        // If it's a number, push it to the output directly
+        if (isdigit(token[0])) {
+            strcat(output,token);
+            strcat(output," ");
+
+            token = strtok(NULL," ");
+            continue;
+        }
+
+        switch (token[0]) {
+            // If it's a left parenthesis, push it to the operator stack for later matching
+            case '(': {
+                operatorStack[stackPointer++] = token[0];
+                break;
+            }
+
+            // If it's a right parenthesis, search for a left parenthesis to match it
+            case ')': {
+                while (operatorStack[stackPointer - 1] != '(') {
```

Comment:
For some reason, the `clang-tidy` checker on Gitpod is reporting a warning here. 🤔
Weird that there's no warning in the GitHub CI.

Reply:
That's certainly very interesting, my local `clang-tidy` reported nothing (Otherwise I would have fixed it before upload) Lemme guess, it claims `loop variable not updated`? (or something like that, I don't remember the exact wording) It's probably a glitch or something, my `clang-tidy` just won't stop screaming when I tried to make the algorithm not output extra space after the final operator because `condition always true` obviously it's not always true because `stackPointer` is decreased in the middle of the loop, but it insists anyway, and I had to give up. my guess is that `strncat(output,&operatorStack[(stackPointer--) - 1],1);` confused it for whatever reason:thinking:

Reply:
I've opened this in Gitpod myself and it claims `the left operand is a garbage value`, it's actually not without merit though, if someone hypothetically inputs something blatantly wrong like ") 3 + 4 (" there can potentially be a segfault. I've tried adding a guard statement, but somehow it continues to warn me anyway, could use some advice before I commit this change